### PR TITLE
Support SHOW CREATE FUNCTION

### PR DESCRIPTION
**Pull request by caithagoras**
Support syntax, semantic analysis, and execution for SHOW CREATE
FUNCTION. An optional list of parameter types may be specified.

For each matching signature, returns the canonical CREATE FUNCTION
that can be used to create the exact same function, along with
the parameter type list of that function signature so that user
can differentiate among the functions.

```
== RELEASE NOTES ==

General Changes
* Add support for SHOW CREATE FUNCTION.
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 1]
```diff
@@ -478,6 +492,66 @@ protected Node visitShowCreate(ShowCreate node, Void context)
...
+                if (!(function instanceof SqlInvokedFunction)) {
+                    throw new PrestoException(GENERIC_USER_ERROR, "SHOW CREATE FUNCTION only supported for SQL functions");
+                }
+
+                SqlInvokedFunction sqlFunction = (SqlInvokedFunction) function;
+                CreateFunction createFunction = new CreateFunction(
+                        node.getName(),
+                        false,
+                        sqlFunction.getParameters().stream()
+                                .map(parameter -> new SqlParameterDeclaration(new Identifier(parameter.getName(), true), parameter.getType().toString()))
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:526]: Wow so we actually rebuild the create function? I thought we remember it in the DB at creation time and just return it.

File: presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java [hunk 2]
```diff
@@ -73,4 +121,81 @@ public void testDropFunctionInvalidFunctionName()
...
+        assertEquals(rows.getMaterializedRows().get(1).getFields(), ImmutableList.of(createFunctionIntFormatted, parameterTypeInt));
+
+        rows = computeActual("SHOW CREATE FUNCTION testing.common.array_append(array(int), int)");
+        assertEquals(rows.getRowCount(), 1);
+        assertEquals(rows.getMaterializedRows().get(0).getFields(), ImmutableList.of(createFunctionIntFormatted, parameterTypeInt));
+
+        rows = computeActual("SHOW CREATE FUNCTION testing.common.rand()");
+        assertEquals(rows.getMaterializedRows().get(0).getFields(), ImmutableList.of(createFunctionRandFormatted, ""));
+
+        assertQueryFails("SHOW CREATE FUNCTION testing.common.array_append()", "Function not found: testing\\.common\\.array_append\\(\\)");
```
Comment (kaikalur) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java:199]: Also add test for things like 'ARRAY_AGG'. What happens when someone tries that?

**Comment by caithagoras** [comment]
[test facebook]

File: presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java (hunk 2 above)
Comment (caithagoras) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestSqlFunctions.java:199]: test added

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 1 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:526]: Yes, we're storing different component of a SQL function as separate columns in the DB, as opposed to the entire `CREATE FUNCTION` statement.

**Comment by caithagoras** [comment]
@rongrong Tests are passing. PR starts from commit 3 "Unify QualifiedName to QualifiedFunctionName conversion".

**Comment by caithagoras** [comment]
[test facebook]

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 3]
```diff
@@ -478,6 +492,66 @@ protected Node visitShowCreate(ShowCreate node, Void context)
             throw new UnsupportedOperationException("SHOW CREATE only supported for tables and views");
         }
 
+        @Override
+        protected Node visitShowCreateFunction(ShowCreateFunction node, Void context)
+        {
+            QualifiedFunctionName functionName = qualifyFunctionName(node.getName());
+            Collection<? extends SqlFunction> functions = metadata.getFunctionManager().getFunctions(session.getRequiredTransactionId(), functionName);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:499]: Hmm, isn't it a little extreme to require transaction to run SHOW CREATE FUNCTION? It's probably ok to just fetch some version?

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:499]: If you want to require transaction for sql functions, make the error message explicit at least.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 4]
```diff
@@ -478,6 +492,66 @@ protected Node visitShowCreate(ShowCreate node, Void context)
...
+                List<TypeSignature> parameterTypes = node.getParameterTypes().get().stream()
+                        .map(TypeSignature::parseTypeSignature)
+                        .collect(toImmutableList());
+                functions = functions.stream()
+                        .filter(function -> function.getSignature().getArgumentTypes().equals(parameterTypes))
+                        .collect(toImmutableList());
+            }
+            if (functions.isEmpty()) {
+                String types = node.getParameterTypes().isPresent()
+                        ? format("(%s)", Joiner.on(", ").join(node.getParameterTypes().get()))
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:510]: you can probably just do `node.getParameterTypes().map(...)`

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 5]
```diff
@@ -478,6 +492,66 @@ protected Node visitShowCreate(ShowCreate node, Void context)
...
+                        : "";
+                throw new PrestoException(FUNCTION_NOT_FOUND, format("Function not found: %s%s", functionName, types));
+            }
+
+            ImmutableList.Builder<Expression> rows = ImmutableList.builder();
+            for (SqlFunction function : functions) {
+                if (!(function instanceof SqlInvokedFunction)) {
+                    throw new PrestoException(GENERIC_USER_ERROR, "SHOW CREATE FUNCTION only supported for SQL functions");
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:518]: *is* only supported

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 6]
```diff
@@ -2186,11 +2193,14 @@ private static boolean isValidUnicodeEscape(char c)
         throw new IllegalArgumentException("Unsupported quantifier: " + symbol.getText());
     }
 
-    private List<String> getTypes(SqlBaseParser.TypesContext types)
+    private Optional<List<String>> getTypes(@Nullable SqlBaseParser.TypesContext types)
```
Comment (rongrong) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:2196]: I like the old way better. It's a bit tedious but it's explicit.

File: presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java [hunk 7]
```diff
@@ -63,6 +63,7 @@
     QUERY_HAS_TOO_MANY_STAGES(0x0000_0028, USER_ERROR),
     INVALID_SPATIAL_PARTITIONING(0x0000_0029, USER_ERROR),
     INVALID_ANALYZE_PROPERTY(0x0000_002A, USER_ERROR),
+    INVALID_FUNCTION_NAME(0x0000_002B, USER_ERROR),
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java:66]: Where is this used?

Comment (caithagoras) [reply presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java:66]: accidental change, removed.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 3 above)
Comment (caithagoras) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:499]: @rongrong
For 1: Transaction ID is guaranteed to exist during query execution, even for any single statement query.
For 2: If Transaction ID is missing, we're having an internal error here. The `getRequiredTransaction` method already checks for `transactionId.isPresent()` and will throw `IllegalArgumentException`, which will be converted to a GENERIC_INTERNAL_ERROR.

**Comment by caithagoras** [comment]
@rongrong @kaikalur Gently ping on the PR.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 3 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java:500]: I don't think missing transaction ID should be an illegal argument exception. I also don't think it's necessary to require transaction ID to exist. The fact that it probably always exist is a different matter. I think function namespaces should still work when there's no transaction. You cannot guarantee versioning consistency, but other than that why it should fail altogether?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java [hunk 8]
```diff
@@ -200,6 +200,17 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
                 .collect(toImmutableList());
     }
 
+    public Collection<? extends SqlFunction> getFunctions(TransactionId transactionId, QualifiedFunctionName functionName)
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java:203]: The implementation here also suggested that `TransactionId` does not need to be required. `FunctionNamespaceManager` would happily take an empty transaction handle. So why do you enforce transaction ID as input? It happened to always exists should not be the reason you always require it to exist.

**Pull request merged**